src/herder/TransactionQueue.cpp

```diff
@@ -127,9 +127,15 @@ findBySeq(int64_t seq, TransactionQueue::TimestampedTransactions& transactions,
         return false;
     }
 
-    releaseAssert(seq - firstSeq <= static_cast<int64_t>(transactions.size()));
-    iter = transactions.begin() + (seq - firstSeq);
-    releaseAssert(iter == transactions.end() || iter->mTx->getSeqNum() == seq);
+    iter = std::lower_bound(transactions.begin(), transactions.end(), seq,
+                            [](TransactionQueue::TimestampedTx const& a,
+                               int64_t b) { return a.mTx->getSeqNum() < b; });
+
+    releaseAssert(iter == transactions.end() || iter->mTx->getSeqNum() >= seq);
+    if (iter != transactions.end() && iter->mTx->getSeqNum() != seq)
```

Comment:
This implementation would allow to "fill in the gaps" because of the `minSeqNum` condition, but the implication would be that we should be able to insert transactions "in the middle" of the source account's `mTransactions`, and the code right now does not know how to deal with that condition.

I think we should update CAP21 (overlay section) to not allow to fill in gaps when receiving txs out of order, and only allow `mTransactions` to grow "from the tail" (it's actually implied right now as the sequence is valid with gaps, so no need to bother to add intermediate txs).

Then we don't have to do binary search either as we would only add txs at the end

Reply:
this points to gaps in tests as this implementation adds "out of order" transactions to `mTransactions` -> breaks subsequent calls to `lower_bound`

Reply:
Correction: I think we still need to do binary search to determine if the transaction is a duplicate

Reply:
`findBySeq` is just used to search for an existing tx right? The validation to prevent filling in the gap is currently done on line 275 when `checkValid` is called with `seqNum` (there's probably a bug here related to fee bumps that I need to look into like you mentioned). The tx seqNum still needs to be greater than the `seqNum` passed into `checkValid` even if `minSeqNum` is set. I wrote a test to confirm this. Am I missing something here?

I agree that the "Transaction forwarding and ordering" section of the CAP should be updated to specify the gap can't be filled. I'll make sure to do this and post the PR here.

Reply:
Oh I see, I now realize that it's just that this function `findBySeq` has a weird return value:
* it returns `false` for txs that are obvious "bad seq" (to avoid calling `checkValid`). This is an optimization.
* it returns `true` otherwise, and the actual result is in `iter` ; `iter == transactions.end()` happens when attempting to add at the end of the transaction queue (in master).

I think we should keep the same semantics and not return `true` as per my comment above.

The contract could be updated to make it clearer that the actual result is the iterator, and the bool is optional and there fast fail.